src/main_nep/nep4.cu

```diff
@@ -223,6 +228,43 @@ static void __global__ find_max_min(const int N, const float* g_q, float* g_q_sc
   }
 }
 
+static __global__ void apply_gnn(
+  const int N,
+  const NEP4::ParaMB paramb,
+  const NEP4::ANN annmb,
+  const NEP4::GNN gnnmb,
+  const float* __restrict__ g_descriptors,
+  const float* __restrict__ g_q_scaler,
+  const int* g_NN,
+  const int* g_NL,
+  float* gnn_descriptors)
+{
+  int n1 = threadIdx.x + blockIdx.x * blockDim.x;
+  if (n1 < N) {
+    int neighbor_number = g_NN[n1]; // Number of neighbors to atom i
+    // get descriptors for atom i and neighbors
+    float q_i[MAX_DIM] = {0.0f};
+    float q_j[MAX_NEIGHBORS * MAX_DIM] = {0.0f}; // maximum size when all atoms are neighbors
+    for (int d = 0; d < annmb.dim; ++d) {
+      q_i[d] = g_descriptors[n1 + d * N] * g_q_scaler[d];
+    }
+    for (int j = 0; j < neighbor_number; ++j) {
+      int n2 = g_NL[n1 + N * j];
+      for (int d = 0; d < annmb.dim; ++d) {
+        q_j[j + d * neighbor_number] = g_descriptors[n2 + d * N] * g_q_scaler[d];
```

Comment:
to save computation, you can perform the normalization using a separate kernel before hand.

Reply:
You mean normalize g_descriptors before the GNN layer?

Reply:
yes, write a separate kernel to do it. In that way, you saved some computations as you don't need to do it for the neighbors.

Reply:
Ah right, smart!